backend/src/services/calculator/per-ingredient-ranker/per-ingredient-ranker.ts

```diff
@@ -0,0 +1,36 @@
+/**
+ * Copyright 2024 Sleep API Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { OptimalTeamSolution } from '@src/domain/combination/combination';
+import { CustomPokemonCombinationWithProduce } from '@src/domain/combination/custom';
+import { calculateSetCover } from '@src/services/calculator/set-cover/calculate-set-cover';
+import { IngredientSet } from 'sleepapi-common';
+import { Ingredient } from 'sleepapi-common/dist/domain/ingredient';
+
+export function rankByIngredient(params: {
+  ingredient: Ingredient;
+  cache: Map<string, CustomPokemonCombinationWithProduce[][]>;
+  reverseIndex: Map<string, CustomPokemonCombinationWithProduce[]>;
+  timeout?: number;
+}): OptimalTeamSolution[] {
+  const singletonIngredient: IngredientSet[] = [{ ingredient: params.ingredient, amount: 1 }];
```

Comment:
very minor: but I think singleton generally represents a single instanced and re-used object of a class. In this case it's just a recipe with 1 quantity used as parameter. Perhaps we should call it "singleIngredientRecipe" or something

Reply:
Changed to `ingAsRecipe` (in the new location).